tests/test_mutations.py

```diff
@@ -472,45 +779,308 @@ def test_keep_mutation_parent_zero_rate(self):
+
+    def pick_allele(self):
+        u = self.rng.flat(0, 1)
+        j = 0
+        while u > self.alphabet_distribution[j]:
+            u -= self.alphabet_distribution[j]
+            j += 1
+        return j
+
+    def transition_allele(self, j):
+        u = self.rng.flat(0, 1)
+        k = 0
+        while u > self.mutation_matrix[j][k]:
+            u -= self.mutation_matrix[j][k]
+            k += 1
+        return k
+
+    def choose_alleles(self, tree_parent, site, mutation_id):
+        if site.new:
+            ai = self.pick_allele()
+            site.ancestral_state = self.alleles[ai]
+        else:
+            ai = self.alleles.index(site.ancestral_state)
+        # sort mutations by (increasing id if both are not null,
+        #  decreasing time, increasing insertion order)
+        print([s.id for s in site.mutations])
+        site.mutations.sort(key=functools.cmp_to_key(cmp_mutation))
+        print([s.id for s in site.mutations])
+        bottom_mutation = {}
+        j = 0
+        while j < len(site.mutations):
+            mut = site.mutations[j]
+            mut.id = mutation_id + j
+            # Traverse up the tree to find the parent mutation
+            # bottom_mutation[u] is the index in mutations of the most recent
+            #    mutation seen on the edge above u so far, if any
+            u = mut.node
+            while u != tskit.NULL and u not in bottom_mutation:
+                u = tree_parent[u]
+            if u == tskit.NULL:
+                # parent is the ancestral state
+                pa = ai
+                assert mut.parent == tskit.NULL
+            else:
+                assert u in bottom_mutation
+                parent_mut = site.mutations[bottom_mutation[u]]
+                mut.parent = parent_mut.id
+                assert mut.parent < mut.id, "Parent after child mutation."
+                assert mut.time <= parent_mut.time, "Parent after child mutation."
+                if mut.new:
+                    pa = self.alleles.index(parent_mut.derived_state)
+                else:
+                    # check that no kept mutation occurs after a new mutation
+                    if parent_mut.new:
+                        raise ValueError("Generated mutation appears above "
+                                         "an existing mutation.")
+
+            keep_mutation = True
+            if mut.new:
+                da = self.transition_allele(pa)
+                if da == pa:
+                    del site.mutations[j]
+                    keep_mutation = False
+                else:
+                    mut.derived_state = self.alleles[da]
+            if keep_mutation:
+                bottom_mutation[mut.node] = j
+                j += 1
+
+    def apply_mutations(self, tables):
+        ts = tables.tree_sequence()
+        positions = sorted(self.sites.keys())
+        j = 0
+        mutation_id = 0
+        tree_parent = np.repeat(tskit.NULL, tables.nodes.num_rows)
```

Comment:
Best to insert a ``dtype=int`` here, numpy can might decide to allocate a double array otherwise which could have unexpected consuquences.

Reply:
`np.repeat( )` doesn't take a `dtype` argument?  But
```
np.array(tskit.NULL).dtype
# dtype('int64')
```

Reply:
OK, looks good. I'm suspicious of numpy's type inference as it can be very quirky. This is probably safe though.